src/Spatial/Euclidean/StraightLine2D.cs

```diff
@@ -0,0 +1,315 @@
+        /// </summary>
+        /// <param name="line">The line to intersect with</param>
+        /// <returns>The point of intersection or null if the lines are parallel</returns>
+        public Point2D? Intersection(StraightLine2D line)
+        {
+            if (IsParallel(line))
+            {
+                return null;
+            }
+
+            if (double.IsInfinity(Gradient))
+            {
+                return new Point2D(xoffset, (line.Gradient * xoffset) + line.yoffset);
+            }
+            else if (double.IsInfinity(line.Gradient))
+            {
+                return new Point2D(line.xoffset, (Gradient * line.xoffset) + yoffset);
+            }
+
+            var denominator = Gradient - line.Gradient;
+
+            var x = (line.yoffset - yoffset) / denominator;
+            var y = (Gradient * x) + yoffset;
+            return new Point2D(x, y);
+        }
+
+        /// <summary>
+        /// Returns the closest point on the line to the given point.
+        /// </summary>
+        /// <param name="p">The point that the returned point is the closest point on the line to</param>
+        /// <returns></returns>
+        public Point2D ClosestPointTo(Point2D p)
+        {
+            StraightLine2D line = new StraightLine2D(p, Gradient);
+            var perpline = line.PerpendicularAt(p);
+            return this.Intersection(perpline).Value;
+        }
+
+        /// <summary>
+        /// Determines if a given point is on this line
+        /// </summary>
+        /// <param name="point">The point to be checked</param>
+        /// <returns>True if the point is on this line</returns>
+        public bool IsThrough(Point2D point)
+        {
+            return IsThrough(point.X, point.Y);
+        }
+
+        /// <summary>
+        /// Determines if a set of coordinates (x,y) is on this line
+        /// </summary>
+        /// <param name="x">the x coordinate</param>
+        /// <param name="y">the y coordinate</param>
+        /// <returns>True if the point is on the line</returns>
+        public bool IsThrough(double x, double y)
+        {
+            return y == (Gradient * x) + yoffset;
+        }
+
+        /// <summary>
+        /// Returns the list of points which lie on a vertical line at coordinate x
+        /// </summary>
+        /// <param name="x">the x coordinate</param>
+        /// <returns>A list of points</returns>
+        public IEnumerable<Point2D> XIntercept(double x = 0)
+        {
+            return new List<Point2D>() { new Point2D(x, (Gradient * x) + yoffset) };
+        }
+
+        /// <summary>
+        /// Returns the list of points which lie on a horizontal line at coordinate y
+        /// </summary>
+        /// <param name="y">the y coordinate</param>
+        /// <returns>A list of points</returns>
+        public IEnumerable<Point2D> YIntercept(double y = 0)
+        {
+            return new List<Point2D>() { new Point2D((y - yoffset) / Gradient, y) };
```

Comment:
I prefer `return new Point2D>[] { new Point2D((y - yoffset) / Gradient, y) };` here but not important.

Reply:
I'm ok with changing that, a fixed array of items is perhaps better in this case than the enumerable one

Reply:
`IReadOnlyList<Point2D>` is nice, shows that there is no lazy stuff going on. And readonly.